File: leetcode/121.py
# Best Time to Buy and Sell Stock

# Time Limit Exceeded
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0
        for i in range(len(prices)-1):
            for j in range(i+1, len(prices)):
                if profit < prices[j]-prices[i]:
                    profit = prices[j]-prices[i]
        return profit


# Passed
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        profit = 0

        for i in range(len(prices)):
            min_price = min(min_price, prices[i])
            profit = max(profit, prices[i]-min_price)

        return profit

File: leetcode/121.go
// Best Time to Buy and Sell Stock

package main

func maxProfit(prices []int) int {
	profit := 0
	for i := 0; i < len(prices)-1; i++ {
		for j := i + 1; j < len(prices); j++ {
			if profit < prices[j]-prices[i] {
				profit = prices[j] - prices[i]
			}
		}
	}
	return profit
}
